src/libraries/System.Net.Security/src/System/Net/Security/TlsFrameHelper.cs

```diff
@@ -235,5 +333,473 @@ public static byte[] CreateAlertFrame(SslProtocols version, TlsAlertDescription
+
+            if (extensionData.Length != supportedVersionLength)
+            {
+                return false;
+            }
+
+            // Get list of protocols we support.I nore the rest.
+            while (extensionData.Length >= VersionLength)
+            {
+                if (extensionData[ProtocolVersionMajorOffset] == ProtocolVersionTlsMajorValue)
+                {
+                    protocols |= TlsMinorVersionToProtocol(extensionData[ProtocolVersionMinorOffset]);
+                }
+
+                extensionData = extensionData.Slice(VersionLength);
+            }
+
+            return true;
+        }
+
+        private static bool TryGetApplicationProtocolsFromExtension(ReadOnlySpan<byte> extensionData, out ApplicationProtocolInfo alpn)
+        {
+            // https://tools.ietf.org/html/rfc7301#section-3.1
+            // opaque ProtocolName<1..2 ^ 8 - 1 >;
+            //
+            // struct {
+            //   ProtocolName protocol_name_list<2..2^16-1>
+            // }
+            // ProtocolNameList;
+            const int AlpnListLengthOffset = 0;
+            const int AlpnListOffset = AlpnListLengthOffset + sizeof(short);
+
+            alpn = ApplicationProtocolInfo.None;
+
+            if (extensionData.Length < AlpnListOffset)
+            {
+                return false;
+            }
+
+            int AlpnListLength = BinaryPrimitives.ReadUInt16BigEndian(extensionData);
+            ReadOnlySpan<byte> alpnList = extensionData.Slice(AlpnListOffset);
+            if (AlpnListLength != alpnList.Length)
+            {
+                return false;
+            }
+
+            while (!alpnList.IsEmpty)
+            {
+                byte protocolLength = alpnList[0];
+                if (alpnList.Length < protocolLength + 1)
+                {
+                    return false;
+                }
+
+                ReadOnlySpan<byte> protocol = alpnList.Slice(1, protocolLength);
+                if (protocolLength == 2)
+                {
+                    if (protocol.SequenceEqual(SslApplicationProtocol.Http2.Protocol.Span))
+                    {
+                        alpn |= ApplicationProtocolInfo.Http2;
+                    }
+                    else if (protocol.SequenceEqual(SslApplicationProtocol.Http3.Protocol.Span))
+                    {
+                        alpn |= ApplicationProtocolInfo.Http3;
+                    }
+                    else
+                    {
+                        alpn |= ApplicationProtocolInfo.Other;
+                    }
+                }
+                else if (protocolLength == SslApplicationProtocol.Http11.Protocol.Length &&
+                         protocol.SequenceEqual(SslApplicationProtocol.Http11.Protocol.Span))
+                {
+                    alpn |= ApplicationProtocolInfo.Http11;
+                }
+                else
+                {
+                    alpn |= ApplicationProtocolInfo.Other;
```

Comment:
When the extension contains an unrecognized ALPN value (other) how do I go back and get the full parsed list of values? I could understand if that version of the API needed to allocate.

Reply:
I was thinking about having two modes based on ProcessingOptions - basic with bitmask without allocation and full where ALPN would be List<SslApplicationProtocol>.  
Alternatively, one could use extension callback.
I wanted to keep the diagnostic simple so did neither one.